File: DecathlonEvalSCript.R
pacman::p_load(readxl,tidyverse,lubridate,dplyr,ggplot2,zoo,tseries,prophet)
df09 <- read_excel("customer_transactions_sample.xlsx",
                   col_types =
                     c("text", "text", "text",
                       "numeric", "date", "numeric", "numeric",
                       "text"),sheet = "Year 2009-2010")

pacman::p_load(readxl,tidyverse,lubridate)
df10 <- read_excel("customer_transactions_sample.xlsx",
                   col_types =
                     c("text", "text", "text",
                       "numeric", "date", "numeric", "numeric",
                       "text"),sheet = "Year 2010-2011")


df11<-rbind.data.frame(df09,df10)
names(df11)[names(df11) == "Customer ID"] <- "CustomerID"
```

```{r,warning=F}
df11$InvoiceDate <- as.Date(df11$InvoiceDate)
df11_aggregated <- df11 %>%
  group_by(InvoiceDate) %>%
  summarise(total_quantity = sum(Quantity))

ggplot(df11_aggregated, aes(x = InvoiceDate, y = total_quantity)) +
  geom_line() +
  labs(x = "Date", y = "Total Quantity") +
  ggtitle("Aggregated Quantity by Day of all Products across Regions including all transactions")


df11<-df11[complete.cases(df11),]
df11<-df11[df11$Quantity>0,]
df11<-df11[df11$Country %in% "United Kingdom",]
df11 <- df11[grepl("CHRISTMAS|TREE|DECORATIONS|SIGNS|LIGHTS|ORNAMENTS", df11$Description, ignore.case = TRUE), ]
a<-aggregate(df11,df11$Quantity~df11$`CustomerID`,sum)
Q1 <- quantile(a$`df11$Quantity`, 0.25)
Q3 <- quantile(a$`df11$Quantity`, 0.75)
IQR <- Q3 - Q1
lower_bound <- Q1 - 1.5 * IQR
upper_bound <- Q3 + 1.5 * IQR
a <- a[a$`df11$Quantity` >= lower_bound & a$`df11$Quantity` <= upper_bound,]
df11<-df11[df11$CustomerID %in% a$`df11$CustomerID`,c("InvoiceDate","Quantity")]
df11$InvoiceDate <- as.Date(df11$InvoiceDate)
df11_aggregated <- df11 %>%
  group_by(InvoiceDate) %>%
  summarise(Quantity_quantity = sum(Quantity))
ggplot(df11_aggregated, aes(x = InvoiceDate, y = Quantity_quantity)) +
  geom_line() +
  labs(x = "Date", y = "Quantity Quantity") +
  ggtitle("Aggregated Quantity by Day")

t1<-ts(df11_aggregated$Quantity_quantity,start = c(2009,12,01),end = c(2011,12,09),frequency = 365)
colnames(df11_aggregated)<-c("ds","y")
t<-df11_aggregated
m<-prophet(t,yearly.seasonality = T,weekly.seasonality = T)
f<-make_future_dataframe(m,periods = 364)
p<-predict(m,f)
pred<-data.frame(yhat=p$yhat,upper=p$yhat_upper,lower=p$yhat_lower)
act<-t$y
act[601:965]<-mean(act)
dates <- seq(as.Date("2009-12-01"), as.Date("2012-12-30"), by = "day")
dates <- dates[weekdays(dates) != "Friday"]
data<-data.frame(act,pred)
data$index<-dates
p <- ggplot(data) +
  geom_line(aes(x = index, y = act, color = "Actual"), size = 1) +
  geom_line(aes(x = index, y = yhat, color = "Predicted"), size = 1) +
  geom_point(aes(x = index, y = upper), color = "grey", size = 0.5) +  
  geom_point(aes(x = index, y = lower), color = "grey", size = 0.5) +  
  scale_color_manual(values = c("Actual" = "red", "Predicted" = "blue")) +
  labs(x = "Date", y = "Values", title = "Actual vs. Predicted with Upper and Lower Bounds") +
  theme_minimal() +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(title = "Data Type"))
print(p)

df11<-rbind.data.frame(df09,df10)
names(df11)[names(df11) == "Customer ID"] <- "CustomerID"
df11$total<-df11$Quantity*df11$Price

df_profit<-df11 %>% group_by(Country) %>% summarise(Profit=sum(Price,na.rm = T))

df_positive <- df11 %>% filter(Quantity > 0)
country_sales <- df_positive %>% group_by(Country) %>% 
  summarise(Quantity_Sold = sum(Quantity, na.rm = TRUE))

df_negative <- df11 %>% filter(Quantity < 0)
country_returns <- df_negative %>% group_by(Country) %>%
  summarise(Quantity_Returned = sum(Quantity, na.rm = TRUE))

country_totals <- merge(country_sales, country_returns, by = "Country", all = TRUE)
country_totals$Quantity_Returned<-abs(country_totals$Quantity_Returned)
country_totals$Quantity_Returned[is.na(country_totals$Quantity_Returned)] <- 0
country_totals$returnrate<-round(country_totals$Quantity_Returned/country_totals$Quantity_Sold*100)

dfhol<-df11
dfhol<-data.frame(day=as.Date(dfhol$InvoiceDate),Country=dfhol$Country)
dfhol<-unique(dfhol)
dfhol<-table(dfhol$Country)
dfhol<-data.frame(dfhol)
colnames(dfhol)<-c("Country","Working_Days")
df12<-merge(dfhol,country_totals)
df13<-merge(df12,df_profit,"Country")
customer_counts<-df11[,c("Country","CustomerID")]
customer_counts<-unique(customer_counts)
customer_counts<-table(customer_counts$Country)
customer_counts<-data.frame(customer_counts)
colnames(customer_counts)<-c("Country","No.Customers")
df14<-merge(df13,customer_counts,"Country")

stockcount<-df11[,c("StockCode","Country")]
stockcount<-table(stockcount$Country)
stockcount<-data.frame(stockcount)
colnames(stockcount)<-c("Country","UniqueProd")
df15<-merge(df14,stockcount,"Country")

data<-df15

lm(formula = Profit~Working_Days,df15)
lm(formula = UniqueProd~No.Customers,df15)

df11$total<-df11$Price*df11$Quantity
names(df11)[names(df11) == "Customer ID"] <- "CustomerID"
df11$InvoiceDate <- as.Date(df11$InvoiceDate)
df_oct_to_dec <- df11 %>%
  filter(month(InvoiceDate) %in% 10:12)
df_rest <- df11 %>%
  filter(!(month(InvoiceDate) %in% 10:12))
head(df_oct_to_dec)
head(df_rest)


sum(df_oct_to_dec$total)/sum(df_rest$total)

df_oct_to_dec<-df_oct_to_dec %>% group_by(Country) %>% summarise(Holiday_revenue=sum(total))
df_rest<-df_rest %>% group_by(Country) %>% summarise(NonHoliday_revenue=sum(total))

df16<-merge(df15,df_rest,"Country")
df16<-merge(df16,df_oct_to_dec,"Country")
df16$holidayrevenueshare<-(df16$Holiday_revenue/df16$NonHoliday_revenue)*100

File: DecathlonEvalSCript.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.formula.api import ols

# Load data
df09 = pd.read_excel("customer_transactions_sample.xlsx", sheet_name="Year 2009-2010")
df10 = pd.read_excel("customer_transactions_sample.xlsx", sheet_name="Year 2010-2011")

# Combine dataframes
df11 = pd.concat([df09, df10])
df11.rename(columns={"Customer ID": "CustomerID"}, inplace=True)

# Data cleaning
df11['InvoiceDate'] = pd.to_datetime(df11['InvoiceDate'])
df11.dropna(subset=['Quantity'], inplace=True)
df11 = df11[df11['Quantity'] > 0]
df11 = df11[df11['Country'] == 'United Kingdom']
df11 = df11[df11['Description'].str.contains(r'CHRISTMAS|TREE|DECORATIONS|SIGNS|LIGHTS|ORNAMENTS', regex=True, case=False)]

# Aggregate data
df11_agg = df11.groupby('CustomerID')['Quantity'].sum().reset_index()

# Remove outliers
Q1 = df11_agg['Quantity'].quantile(0.25)
Q3 = df11_agg['Quantity'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
df11_agg = df11_agg[(df11_agg['Quantity'] >= lower_bound) & (df11_agg['Quantity'] <= upper_bound)]

# Plot aggregated quantity by day
df11_aggregated = df11.groupby('InvoiceDate')['Quantity'].sum().reset_index()
plt.figure(figsize=(10, 6))
sns.lineplot(data=df11_aggregated, x='InvoiceDate', y='Quantity')
plt.xlabel('Date')
plt.ylabel('Total Quantity')
plt.title('Aggregated Quantity by Day')
plt.show()

# Prophet forecasting
from fbprophet import Prophet

df11_aggregated.columns = ['ds', 'y']
m = Prophet(yearly_seasonality=True, weekly_seasonality=True)
m.fit(df11_aggregated)
future = m.make_future_dataframe(periods=364)
forecast = m.predict(future)

# Plot actual vs. predicted with upper and lower bounds
plt.figure(figsize=(10, 6))
plt.plot(df11_aggregated['ds'], df11_aggregated['y'], color='red', label='Actual', linewidth=1)
plt.plot(forecast['ds'], forecast['yhat'], color='blue', label='Predicted', linewidth=1)
plt.fill_between(forecast['ds'], forecast['yhat_upper'], forecast['yhat_lower'], color='grey', alpha=0.3)
plt.xlabel('Date')
plt.ylabel('Values')
plt.title('Actual vs. Predicted with Upper and Lower Bounds')
plt.legend()
plt.show()

# Calculate profit and sales
df11['Total'] = df11['Price'] * df11['Quantity']
df_profit = df11.groupby('Country')['Price'].sum().reset_index()

df_positive = df11[df11['Quantity'] > 0]
country_sales = df_positive.groupby('Country')['Quantity'].sum().reset_index()

df_negative = df11[df11['Quantity'] < 0]
country_returns = df_negative.groupby('Country')['Quantity'].sum().reset_index()
country_returns['Quantity'] = abs(country_returns['Quantity'])

country_totals = pd.merge(country_sales, country_returns, on='Country', how='outer')
country_totals.fillna(0, inplace=True)
country_totals['returnrate'] = round((country_totals['Quantity_y'] / country_totals['Quantity_x']) * 100, 2)

# Load holiday data
df_hol = df11[['InvoiceDate', 'Country']]
df_hol = df_hol.drop_duplicates()
df_hol = df_hol.groupby('Country').size().reset_index(name='Working_Days')

# Merge all data
df16 = pd.merge(country_totals, df_profit, on='Country')
df16 = pd.merge(df16, df_hol, on='Country')

# Calculate holiday revenue share
df16['holidayrevenueshare'] = (df16['Holiday_revenue'] / df16['NonHoliday_revenue']) * 100

# Plot country by holiday revenue share
plt.figure(figsize=(10, 6))
sns.barplot(data=df16.sort_values(by='holidayrevenueshare', ascending=False),
            x='Country', y='holidayrevenueshare',
            palette={'United Kingdom': 'darkblue', 'Others': 'skyblue'})
plt.xticks(rotation=45, ha='right')
plt.ylabel('Holiday Revenue Share (%)')
plt.title('Country by Holiday Revenue Share')
plt.tight_layout()
plt.show()
